### PR TITLE
Datastore len check fix

### PR DESCRIPTION
**Pull request by NishanthJKumar**
Adds the datastore length check back to `_try_specializing_pnad` method for backchaining. Also adds a test that distinguishes this change.

### CODE REVIEW

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 1]
```diff
@@ -283,6 +287,13 @@ def _try_specializing_pnad(
         new_pnad = self._create_new_pnad_with_params_and_add_effects(
             pnad, updated_params, updated_add_effects)
 
+        if check_datastore_change:
+            # If the new PNAD has a datastore size that's not the same
+            # as that of the original PNAD, then we've potentially lost some
+            # data by specializing, which might do harm!
+            if len(new_pnad.datastore) != len(pnad.datastore):
```
Comment (tomsilver) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:294]: for my own understanding, could we assert here that that the lengths are related via < or > if they are not equal?

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 2]
```diff
@@ -133,7 +133,8 @@ def _learn(self) -> List[PartialNSRTAndDatastore]:
                     else:
                         new_pnad = self._try_specializing_pnad(
                             necessary_add_effects,
-                            param_opt_to_general_pnad[option.parent], segment)
+                            param_opt_to_general_pnad[option.parent], segment,
+                            False)
```
Comment (ronuchit) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:137]: use keyword argument here: `check_datastore_change=False`

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 3]
```diff
@@ -237,14 +238,17 @@ def _try_specializing_pnad(
...
+        check_datastore_change: bool = True
     ) -> Optional[PartialNSRTAndDatastore]:
         """Given a PNAD and some necessary add effects that the PNAD must
         achieve, try to make the PNAD's add effects more specific
         ("specialize") so that they cover these necessary add effects.
 
         Returns the new constructed PNAD, without modifying the
         original. If the PNAD does not have a grounding that can even
-        partially satisfy the necessary add effects, returns None.
+        partially satisfy the necessary add effects, or if the newly
```
Comment (ronuchit) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:249]: Fix this docstring -- this second condition only matters if `check_datastore_change` is True.

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py (hunk 1 above)
Comment (ronuchit) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py:294]: +1, i think we can `assert len(new_pnad.datastore) < len(pnad.datastore)` in here, since data can never be added to the datastore?

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 4]
```diff
@@ -283,6 +287,13 @@ def _try_specializing_pnad(
         new_pnad = self._create_new_pnad_with_params_and_add_effects(
             pnad, updated_params, updated_add_effects)
 
+        if check_datastore_change:
+            # If the new PNAD has a datastore size that's not the same
+            # as that of the original PNAD, then we've potentially lost some
+            # data by specializing, which might do harm!
+            if len(new_pnad.datastore) != len(pnad.datastore):
+                return None
+
```
Comment (ronuchit) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:296]: have you verified this PR across several seeds of rnt and rnt single option?

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py [hunk 5]
```diff
@@ -27,7 +31,7 @@ def find_unification(necessary_add_effects,
 
 
 def test_backchaining_strips_learner():
-    """Test the BackchainingSidePredicateLearner."""
+    """Test the BackchainingSTRIPSLearner."""
```
Comment (ronuchit) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:34]: good catch. the word "BackchainingSidePredicateLearner" appears in one other place in this file, can you fix that too?

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py [hunk 6]
```diff
@@ -111,6 +117,134 @@ def test_backchaining_strips_learner():
     assert str(pnads[0]) == repr(pnads[0]) == expected_str
 
 
+def test_backchaining_strips_learner_order_dependence():
+    """Test that the BackchainingSTRIPSLearner is invariant to order of
+    traversal through trajectories."""
+    # Setup up the types and predicates.
```
Comment (ronuchit) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:123]: Set up*

(also, this appears in behavior_options.py, so if you want, you can update "Setup" -> "Set up" there too)

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py [hunk 7]
```diff
@@ -111,6 +117,134 @@ def test_backchaining_strips_learner():
...
+    learner = MockBackchainingSTRIPSLearner(
+        [traj3, traj2, traj1], [task1, task2, task3],
+        {RobotAt, LightOn, NotLightOn, LightColorBlue, LightColorRed},
+        [[segment3], [segment2], [segment1]])
+    reverse_order_pnads = learner.learn()
+
+    # First, check that the two sets of PNADs have the same number of PNADs.
+    assert len(natural_order_pnads) == len(reverse_order_pnads) == 2
+
+    correct_pnads = [
```
Comment (ronuchit) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:219]: can this be a set, or does order matter?

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py [hunk 8]
```diff
@@ -86,17 +90,19 @@ def test_backchaining_strips_learner():
 
     # Test sidelining where an existing operator needs to
     # be specialized.
-    goal3 = {Asleep([bob]), Sad([bob])}
```
Comment (ronuchit) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:89]: just curious what's the motivation behind the changes to this existing test?

**Review by ronuchit — COMMENTED** [review]
i love seeing 100 lines of tests for a 10 line fix! :D just some minor comments

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py (hunk 8 above)
Comment (NishanthJKumar) [reply tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:89]: that test turned out to be wrong in that it was relying on the bug (i.e, not checking datastore length change) to pass. I modified the input so that it would produce the same output given our fix.

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py (hunk 4 above)
Comment (NishanthJKumar) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py:296]: good call - just verified results on both envs across 5 seeds! Things work as expected.

**Review by ronuchit — APPROVED** [review]
